### DISCUSSION TITLE
How to verify that dev dependencies are only used in tests?

### DISCUSSION THREAD

**Discussion by zkochan** [pnpm/pnpm#3851] [General]
In the past I developed [package-preview](https://github.com/zkochan/package-preview) but it doesn't work with monorepos. We should somehow avoid issues like https://github.com/pnpm/pnpm/pull/3850

**Comment by gaggle** [comment]
Back when I wrote Python I did a two-pass test-automation run:
1. First run the test-suite with all dependencies installed, gathering code coverage etc., just like normal.
2. Build and pack the package and only install prod dependencies, and then run the test-suite again targeting that package.

I can't recall the exact mechanism I exploited for this (maybe manipulating PYTHON_PATHS and making sure tests always referenced package-code via the package-name, never via relative files). But it was a nice confidence-booster to see tests be green even when all dev-dependencies were gone.

I can't connect the dots of how to do this in JS though. I asked on Twitter earlier this year and got a few promising responses, but nothing I've followed up on yet: https://twitter.com/jonlauridsen/status/1357264592137773060.

Anyway, just ideas, it's an interesting topic.
